crates/core/src/db/datastore/traits.rs

```diff
@@ -498,3 +498,32 @@ pub trait MutTxDatastore: TxDatastore + MutTx {
         row: ProductValue,
     ) -> Result<ProductValue>;
 }
+
+/// Describes a programmable [`TxDatastore`].
+///
+/// A programmable datastore is one which has a program of some kind associated
+/// with it.
+pub trait Programmable: TxDatastore {
+    /// Retrieve the [`Hash`] of the program currently associated with the
+    /// datastore.
+    ///
+    /// A `None` result means that no program is currently associated, e.g.
+    /// because the datastore has not been fully initialized yet.
+    fn program_hash(&self, tx: &Self::TxId) -> Result<Option<Hash>>;
+}
+
+/// Describes a [`Programmable`] datastore which allows to update the program
+/// associated with it.
+pub trait MutProgrammable: MutTxDatastore {
+    /// A fencing token (usually a monotonic counter) which allows to order
+    /// `set_module_hash` with respect to a distributed locking service.
+    type FencingToken: Eq + Ord;
+
+    /// Update the [`Hash`] of the program currently associated with the
+    /// datastore.
+    ///
+    /// The operation runs within the transactional context `tx`. The fencing
+    /// token `fence` must be verified to be greater than in any previous
+    /// invocations of this method.
+    fn set_program_hash(&self, tx: &mut Self::MutTxId, fence: Self::FencingToken, hash: Hash) -> Result<()>;
```

Comment:
Does this cause the program itself to be updated, or only the stored hash?

Reply:
Only the hash — but inside the transaction which updates the module.

Reply:
It's a good point, actually -- the datastore doesn't store its own module. That does make sense in a distributed setting, but is another piece to worry about for keeping a replica consistent. Food for thought...